### PR TITLE
BE: User tools

### PR DESCRIPTION
**Pull request by EugeneLightsOn**
- Adding tools to the User model
- Extending the API
- User tools take priority over agent tools
**AI Description**

This PR introduces the ability for users to have associated tools, enhancing the user model and impacting various backend components.

**Code Changes:**
- Adds a `tools` attribute to the `User` model, allowing users to have a list of associated tools.
- Updates the UserBase schema to include an optional `tools` attribute.
- Introduces a new function, `get_user_by_email`, in the user.py module, enabling retrieval of a user by their email address.
- Enhances the create_user and update_user functions in user.py with transaction validation.
- Modifies the create_user and update_user functions in routers/user.py to include request validation dependencies.
- Extends the create_user and update_user tests to cover scenarios with and without tools.
- Updates the process_chat function in services/chat.py to incorporate user tools and introduces new functions for processing user and agent tools.
- Adds a new function, validate_user_request, in services/request_validators.py to validate create user requests.
- Includes a new module, services/user.py, with a function to validate the existence of a user.
- Adjusts the User factory in tests/unit/factories/user.py to include a `tools` attribute.

**Reasoning:**
The primary purpose of this PR is to enable users to have associated tools, enhancing the user model and related functionality. This change involves modifying the `User` model to include a `tools` attribute, represented as a list of strings. This addition allows users to have a customizable set of tools associated with their account.

To support this new feature, various backend components have been updated. The `create_user` and `update_user` functions have been enhanced with transaction validation, ensuring data integrity during user creation and modification. Request validation has also been introduced to the corresponding router functions, providing additional input validation for user creation and updates.

The testing infrastructure has been expanded to cover scenarios involving user tools. This includes creating users with specific tools and updating user toolsets, ensuring that the new functionality behaves as expected.

Additionally, the `process_chat` function has been refactored to incorporate user tools, introducing new functions to handle tool processing for both users and agents. This change allows user tools to influence the chat process, potentially customizing the chat experience based on the user's associated tools.

Overall, this PR introduces a significant enhancement to the user model and related functionality, providing users with the ability to have and utilize associated tools.

### CODE REVIEW

File: src/backend/crud/user.py [hunk 1]
```diff
@@ -50,6 +52,21 @@ def get_users(db: Session, offset: int = 0, limit: int = 100) -> list[User]:
     return db.query(User).order_by(User.fullname).offset(offset).limit(limit).all()
 
 
+def get_user_by_email(db: Session, email: str) -> User:
```
Comment (tianjing-li) [inline src/backend/crud/user.py:55]: Should the validate_transaction decorator be added to this?

File: src/backend/services/chat.py [hunk 2]
```diff
@@ -174,6 +158,65 @@ def process_chat(
     )
 
 
+def check_chat_request_tools(
+    chat_request: BaseChatRequest, entity: Agent | User
+) -> None:
+    if chat_request.tools:
+        entity_type = "agent" if isinstance(entity, Agent) else "user"
```
Comment (tianjing-li) [inline src/backend/services/chat.py:165]: Best to explicitly validate in the else. scenario if `entity` is `User`, and throw some exception if its neither Agent|User

File: src/backend/services/chat.py [hunk 3]
```diff
@@ -174,6 +158,65 @@ def process_chat(
     )
 
 
+def check_chat_request_tools(
+    chat_request: BaseChatRequest, entity: Agent | User
+) -> None:
```
Comment (tianjing-li) [inline src/backend/services/chat.py:163]: Let's add a docstring here

File: src/backend/services/chat.py [hunk 4]
```diff
@@ -174,6 +158,65 @@ def process_chat(
...
+        chat_request (BaseChatRequest): Chat request data.
+        ctx (Context): Context object.
+
+    Returns:
+        dict: Dictionary containing the processed tools.
+    """
+    user_id = ctx.get_user_id()
+    agent_id = ctx.get_agent_id()
+    user = None
+    agent = None
```
Comment (tianjing-li) [inline src/backend/services/chat.py:190]: We probably don't need to call `get_user_id()` and `get_agent_id()`, you can shorten it to 
```
ctx.with_user(session)
user = ctx.get_user()
```
and directly check if `user is None`

Comment (tianjing-li) [reply src/backend/services/chat.py:190]: Same thing for agent

File: src/backend/services/request_validators.py [hunk 5]
```diff
@@ -396,3 +396,30 @@ async def validate_organization_request(session: DBSessionDep, request: Request)
         raise HTTPException(
             status_code=400, detail=f"Organization with name: {name} already exists."
         )
+
+
+async def validate_user_request(session: DBSessionDep, request: Request):
```
Comment (tianjing-li) [inline src/backend/services/request_validators.py:401]: If this is to explicitly check create User requests, let's change the method  name to `validate_create_user_request`

File: src/backend/services/user.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+from fastapi import HTTPException
+
+from backend.crud import user as user_crud
+from backend.database_models.database import DBSessionDep
+from backend.database_models.user import User
+
+
+def validate_user_exists(session: DBSessionDep, user_id: str) -> User:
```
Comment (tianjing-li) [inline src/backend/services/user.py:8]: Let's add a docstring

**Review by tianjing-li — CHANGES_REQUESTED** [review]
Couple suggestions

File: src/backend/services/chat.py (hunk 2 above)
Comment (EugeneLightsOn) [reply src/backend/services/chat.py:165]: Thanks

File: src/backend/services/request_validators.py (hunk 5 above)
Comment (EugeneLightsOn) [reply src/backend/services/request_validators.py:401]: Done

File: src/backend/services/user.py (hunk 6 above)
Comment (EugeneLightsOn) [reply src/backend/services/user.py:8]: Thanks

File: src/backend/services/chat.py (hunk 4 above)
Comment (EugeneLightsOn) [reply src/backend/services/chat.py:190]: I need these IDs for the exception details

File: src/backend/services/chat.py (hunk 3 above)
Comment (EugeneLightsOn) [reply src/backend/services/chat.py:159]: Done

File: src/backend/services/chat.py (hunk 2 above)
Comment (EugeneLightsOn) [reply src/backend/services/chat.py]: Thanks

**Comment by EugeneLightsOn** [comment]
> Couple suggestions

Thanks TJ, everything is resolved

**Review by tianjing-li — APPROVED** [review]
looks good to me now

**Pull request merged**